### DISCUSSION TITLE
Doubt about using Redis with Bucket4j

### DISCUSSION THREAD

**Discussion by LMoraes7** [vladimir-bukhtoyarov/bucket4j#278] [Q&A] [answered]
Hello everyone, how are you?

I saw in this issue -> #36, that support for Redis along with Redisson was added in version 7.0.0, but this support has not been documented yet.

I'm still quite new in the world of software development, so any ignorance on this subject on my part, please forgive me. @vladimir-bukhtoyarov or someone who already has experience with this Bucket4j library with redis support, could you confirm me what the purpose of the Duration ttl argument would be? Is the purpose of this argument to indicate how long the bucket created by this ProxyManager should remain stored in the Redis cache?

`
public RedissonBasedProxyManager(CommandExecutor commandExecutor, ClientSideConfig clientSideConfig, Duration ttl) {
        super(clientSideConfig);
        this.commandExecutor = Objects.requireNonNull(commandExecutor);
        this.ttlMillis = ttl.toMillis();
    }
`

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
@LMoraes7 hello.

> Is the purpose of this argument to indicate how long the bucket created by this ProxyManager should remain stored in the Redis cache?

Yes, you are right

**Reply by LMoraes7** [reply]
Is this RedissonBasedProxyManager TTL based on the record creation time or is it based on the last access (every time I access this record the TTL becomes (1 minute + last access time)) to this record?

Example for context: If I create a new bucket for a given IP and in RedissonBasedProxyManager I set the TTL to 1 minute, regardless of how many accesses I make to this new bucket, 1 minute after its creation the bucket will expire in Redis.

**Reply by LMoraes7** [reply]
In some very simple tests that I did here, I noticed that the TTL is behaving as follows: on the last token that is successfully consumed (tryConsume(any long) returning true), the TTL will be added to the value we defined in RedissonBasedProxyManager. From what I understand, the bucket TTL logic is as follows -> (TTL defined in RedissonBasedProxyManager + last time a token was successfully consumed).

**Reply by vladimir-bukhtoyarov** [reply]
Yes. You are right, TTL of bucket is prolonged each time when persisting bucket. In the upcoming release 7.6.0 will be provided more flexible way by ability to configure [ExpirationStrategy](https://github.com/bucket4j/bucket4j/blob/7.6/bucket4j-core/src/main/java/io/github/bucket4j/distributed/ExpirationStrategy.java), it will be 3 built-in implementation as well as will be provided ability to implement own.

**Reply by LMoraes7** [reply]
Thank you @vladimir-bukhtoyarov, I understood about the functioning of TTL, I look forward to this new functionality.

 Changing a bit on the subject of TTL. Reading some issues in your project, I saw that you and other developers did not recommend using Bucket4j with Redisson using the JCache specification due to race condition issues when consuming tokens from the bucket. But from what I understand, this problem occurred when using Redisson through the JCache specification.

 I'm looking to use the bucket4j-redis library you developed in a production application. Is it safe to use this bucket4j-redis in a distributed systems environment?

**Reply by vladimir-bukhtoyarov** [reply]
Direct usage of [RedissonBasedProxyManager](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-redis/src/main/java/io/github/bucket4j/redis/redisson/cas/RedissonBasedProxyManager.java) is safe. Usage through redisson jcache is not safe.
